src/ZoneServer/World/Actors/Characters/Components/CollectionComponent.cs

```diff
@@ -0,0 +1,165 @@
+		/// loaded from the database.
+		/// </remarks>
+		/// <param name="collectionId"></param>
+		/// <param name="registeredItems"></param>
+		public void AddSilent(int collectionId, List<int> registeredItems)
+		{
+			lock (_syncLock)
+			{
+				if (ZoneServer.Instance.Data.CollectionDb.TryFind(collectionId, out var collectionData))
+				{
+					_collectionList.Add(collectionId, collectionData);
+					_collectionProgress.Add(collectionId, registeredItems);
+					if (registeredItems.Count == collectionData.RequiredItems.Count)
+					{
+						// TODO: Add the reward to the list of earned rewards
+					}
+				}
+			}
+		}
+
+		/// <summary>
+		/// Gets a collection and returns the list of items registered to it.
+		/// </summary>
+		/// <param name="collectionId"></param>
+		/// <param name="registeredItems"></param>
+		/// <returns></returns>
+		public bool TryGet(int collectionId, out List<int> registeredItems)
+		{
+			lock (_syncLock)
+				return _collectionProgress.TryGetValue(collectionId, out registeredItems);
+		}
+
+		/// <summary>
+		/// Returns the total number of registered collections
+		/// </summary>
+		/// <returns></returns>
+		public int Count()
+		{
+			lock (_syncLock)
+				return _collectionList.Count;
+		}
+
+		/// <summary>
+		/// Returns the list of all collections the user has registered
+		/// </summary>
+		/// <returns></returns>
+		public List<int> GetList()
+		{
+			lock (_syncLock)
+				return _collectionList.Keys.ToList();
+		}
+
+		/// <summary>
+		/// Returns the list of items currently registered to the given collection
+		/// This list might be empty if no items are registered
+		/// Returns null if the collection isn't registered at all
+		/// </summary>
+		/// <returns></returns>
+		public List<int> GetProgress(int collectionId)
+		{
+			lock (_syncLock)
+			{ 
+				if (_collectionProgress.TryGetValue(collectionId, out var progress))
+					return progress;
+				return null;
+			}
+		}
+
+
+		/// <summary>
+		/// Registers an item to this collection
+		/// <param name="collectionId"></param>
+		/// <param name="item"></param>
+		/// </summary>
+		public bool RegisterItem(int collectionId, Item item)
+		{
+			lock (_syncLock)
+			{
+				if (_collectionList.TryGetValue(collectionId, out var collectionData))
+					if (collectionData.RequiredItems.Contains(item.Data.ClassName))
+						if (_collectionProgress.TryGetValue(collectionId, out var collectionProgress))
+						{ 
+							// First need to check how many times this item appears in the collection
+							var itemsNeeded = collectionData.RequiredItems.Count(a => a.Equals(item.Data.ClassName));
+							var itemsAddedAlready = collectionProgress.Count(a => a == item.Id);
+
+							if (itemsNeeded > itemsAddedAlready)
+							{ 
+								collectionProgress.Add(item.Data.Id);
+								if (collectionProgress.Count == collectionData.RequiredItems.Count)
+								{
+									// TODO: Give the award here and re-calculate stats
```

Comment:
Is there a hurdle with adding the rewards?

Reply:
Is there already a system in place to add bonus stats to a character?  I didn't see it when I looked for it and I felt that creating such a system was outside the scope of this PR.

I think I would prefer to handle this in a separate PR in any case, since the code to add the rewards could be substantial.

Reply:
I'm not familiar with what exactly the rewards are, but granting bonus stats might be as simple as just calling modify on the stat you want to increase, which would fairly trivial.

Reply:
There's a bunch of different reward types.

The most simple are derived stats, like physical attack +5.  This is indeed probably fairly trivial.  The only real issue is that if there's any situation where the game recalculates your derived stats from scratch, we'd have to re-add these bonus stats again (probably needs to be a method like "Collections.AddBonusStats" or something like that that a method could call).

You can also get base stats, like STR and INT.  I notice that in the interface, there are 3 categories shown for these stats, base, investment, and other.  This definitely needs to go in "Other".  Does a method of adding to other currently exist?  (If so, equipment probably also uses it).

The third reward type is items.  This is also pretty easy.  When the collection is finished (at the exact time of entering the final item), it gives you the item(s).  When calling the method to get all of your existing bonuses, these ones are ignored, so it doesn't give you the items over and over.

The fourth type is account properties.  It seems like the only one of these is maximum team storage size.  I'm not sure if this is implemented yet?  Might need a separate method from the others to get this.

In any case, it's complex enough that I feel like we might want to save this for another PR.

Reply:
When it comes to bonuses and rewards I'm of the opinion that a script will usually yield the best results, to make the system flexible for any new official additions, as well as users' customization ideas. The setup is a bit more involved than setting a type and one or two parameters, but in turn you can do pretty much anything with it. Would that seem like a good approach in this case?

I think adding the base system to accomplish granting rewards should ideally be in place before merge, even if not every single reward type is implemented yet.